### PR TITLE
Project 3 Blobstore Library

### PR DESCRIPTION
**Pull request by kinsaurralde**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
+
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Remove empty line

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * On GET request, return url to upload image to
+ * On POST request, upload image and return url where it is stored
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: It seems these should be the the comment of the get and post methods. You can have more general description about this class.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+ */
+@WebServlet("/blobstore")
+public class BlobstoreServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: You can define this as a member variable, similar to the datastore object in another file.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(request, "image");
+    response.setContentType("text/html");
+    response.getWriter().println(imageUrl);
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Same here. You can reuse the member variable.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Maybe change this to the following format:
// TODO: xxx

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: can be member variable as well

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * On GET request, return url to upload image to
+ * On POST request, upload image and return url where it is stored
+ */
+@WebServlet("/blobstore")
+public class BlobstoreServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/blobstore");
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Try to use as less as string literals in functions as possible. If the string is a constants, define it as a constant in the class with "private static final" modifiers.

The same as the one below "text/html"

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    System.out.println("GET " + uploadUrl);
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: It's fine to leave this as is if you find standard output is easier to debug. Usually we just log the message.

**Review by huzcn — CHANGES_REQUESTED** [review]
Nice work!

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java (hunk 5 above)
Comment (kinsaurralde) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: These were comments from walk through code that im not using in the next PR so I will just delete this for now.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 9]
```diff
@@ -77,11 +79,8 @@ private String getUploadedFileUrl(HttpServletRequest request, String formInputEl
       return null;
     }
 
-    // We could check the validity of the file here, e.g. to make sure it's an image file
-    // https://stackoverflow.com/q/10779564/873165
-
     // Use ImagesService to get a URL that points to the uploaded file.
-    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: remove empty line?

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 10]
```diff
@@ -34,31 +33,34 @@
...
+ * Handles image uploads
  */
 @WebServlet("/blobstore")
 public class BlobstoreServlet extends HttpServlet {
+  private final BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+  private final ImagesService imagesService = ImagesServiceFactory.getImagesService();
+  private static final String CONTENT_TYPE_TEXT = "text/html";
+  private static final String UPLOAD_URL = "/blobstore";
+
+  /** Return url image should be uploaded to */
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Returns

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java [hunk 11]
```diff
@@ -34,31 +33,34 @@
...
-    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
-    String uploadUrl = blobstoreService.createUploadUrl("/blobstore");
-    System.out.println("GET " + uploadUrl);
-    response.setContentType("text/html");
+    String uploadUrl = blobstoreService.createUploadUrl(UPLOAD_URL);
+    response.setContentType(CONTENT_TYPE_TEXT);
     response.getWriter().println(uploadUrl);
   }
 
+  /** Return url where image is now stored */
```
Comment (huzcn) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreServlet.java]: Returns

**Review by huzcn — APPROVED** [review]
Nice! Approve for merging after nits are resolved

**Pull request merged**